### PR TITLE
fix(GODT-1578,GODT-1612): mailbox name matching and Noselect estimation.

### PR DESCRIPTION
**Pull request by cuthix**

### CODE REVIEW

File: internal/state/match.go [hunk 1]
```diff
@@ -21,52 +21,78 @@ type Match struct {
...
+			if !ok || // is deleted superior
+				matchedName == "" || // is empty request for delimiter response
+				subscribed && !mbox.Subscribed { // is unsubscribed superior
+				matches[matchedName] = Match{
+					Name:      matchedName,
 					Delimiter: delimiter,
 					Atts:      imap.NewFlagSet(imap.AttrNoSelect),
 				}
+
+				continue
```
Comment (jameshoulahan) [inline internal/state/match.go:71]: could you extract this stuff into some helper functions `skipMatch` and `addMatch`?

File: internal/state/match.go [hunk 2]
```diff
@@ -21,52 +21,78 @@ type Match struct {
...
+			}
```
Comment (jameshoulahan) [inline internal/state/match.go:72]: could you extract this into some helper functions?

File: tests/delete_test.go [hunk 3]
```diff
@@ -14,43 +14,87 @@ func TestDelete(t *testing.T) {
...
+			`* LIST (\Unmarked) "/" "blurdybloop"`,
+			`* LIST (\Marked) "/" "foo"`,
+			`* LIST (\Marked) "/" "foo/bar"`,
+		)
+		c.OK(`A005`)
+
+		c.C(`A006 STATUS "foo/bar" (UIDNEXT MESSAGES)`)
+		c.S(
+			`* STATUS "foo/bar" (UIDNEXT 2 MESSAGES 1)`,
+		)
```
Comment (jameshoulahan) [inline tests/delete_test.go:75]:
```suggestion
		c.S(`* STATUS "foo/bar" (UIDNEXT 2 MESSAGES 1)`)
```

File: tests/delete_test.go [hunk 4]
```diff
@@ -14,43 +14,87 @@ func TestDelete(t *testing.T) {
...
+		c.OK(`A006`)
+
+		c.C(`A007 DELETE "foo"`).OK(`A007`)
+
+		c.C(`A008 STATUS "foo/bar" (UIDNEXT MESSAGES)`)
+		c.S(
+			`* STATUS "foo/bar" (UIDNEXT 2 MESSAGES 1)`,
+		)
```
Comment (jameshoulahan) [inline tests/delete_test.go:83]:
```suggestion
		c.S(`* STATUS "foo/bar" (UIDNEXT 2 MESSAGES 1)`)
```

File: tests/list_test.go [hunk 5]
```diff
@@ -170,17 +168,54 @@ func TestListRemoved(t *testing.T) {
...
+			`* LIST (\Unmarked) "." "some.thing"`,
+			`* LIST (\Unmarked) "." "some.thing.else"`,
+			`* LIST (\Unmarked) "." "some.thing.else.entirely"`,
+		)
+		c.OK(`tag`)
+
+		c.C(`tag list "some.thing" "%"`)
+		c.S(
+			`* LIST (\Unmarked) "." "some.thing"`,
+		)
```
Comment (jameshoulahan) [inline tests/list_test.go:204]:
```suggestion
		c.S(`* LIST (\Unmarked) "." "some.thing"`)
```

File: tests/list_test.go [hunk 6]
```diff
@@ -170,17 +168,54 @@ func TestListRemoved(t *testing.T) {
...
+		c.S(
+			`* LIST (\Unmarked) "." "some.thing.else"`,
+			`* LIST (\Unmarked) "." "some.thing.else.entirely"`,
+		)
+		c.OK(`tag`)
+
+		c.C(`tag list "some.thing." "%"`)
+		c.S(
+			`* LIST (\Unmarked) "." "some.thing.else"`,
+		)
```
Comment (jameshoulahan) [inline tests/list_test.go:217]:
```suggestion
		c.S(`* LIST (\Unmarked) "." "some.thing.else"`)
```

File: tests/lsub_test.go [hunk 7]
```diff
@@ -6,38 +6,93 @@ import (
...
 
-		c.C(`A007 LSUB "#news." "comp.%"`)
-		c.S(`* LSUB (\Noselect) "." "#news.comp.mail"`)
-		c.OK(`A007`)
+		c.C(`A002 LSUB "" "foo"`)
+		// TODO: The server MUST NOT unilaterally remove an existing mailbox name
+		// from the subscription list even if a mailbox by that name no
+		// longer exists.
+		//
+		// c.S(`* LSUB (\Noselect) "." "foo"`)
```
Comment (jameshoulahan) [inline tests/lsub_test.go:95]: do we have an issue created for this to be resolved?

Comment (cuthix) [reply tests/lsub_test.go]: yup added ticket to comment

File: internal/state/match.go (hunk 2 above)
Comment (cuthix) [reply internal/state/match.go]: did you mean like this?

**Review by jameshoulahan — APPROVED** [review]

**Pull request merged**